server/channels/app/authentication.go

```diff
@@ -59,7 +59,24 @@ func (a *App) IsPasswordValid(rctx request.CTX, password string) *model.AppError
 	return nil
 }
 
-func (a *App) CheckPasswordAndAllCriteria(rctx request.CTX, user *model.User, password string, mfaToken string) *model.AppError {
+func (a *App) CheckPasswordAndAllCriteria(rctx request.CTX, u *model.User, password string, mfaToken string) *model.AppError {
+	// MM-37585
+	// The user param will have stale info. in case of concurrent failed login attempts by the same user.
+	// So, lock and fetch the updated attempts count.
+	a.ch.loginAttemptsMut.Lock()
+	defer a.ch.loginAttemptsMut.Unlock()
+
+	a.Srv().Store().User().InvalidateProfileCacheForUser(u.Id)
```

Comment:
Why are we invalidating twice? Here and L84.

Reply:
Will remove both of them. Since, the user cache layer is already handling the invalidation.